subsys/net/lib/config/init_clock_sntp.c

```diff
@@ -112,10 +112,230 @@ int net_init_clock_via_sntp(void)
+		/* If address was found, schedule SNTP query, if that fails schedule next retry. */
+		ret = sntp_query_async(net_sad(&sntp_addr), sntp_addrlen);
+		if (ret < 0) {
+			k_work_reschedule(&sntp_resync_work_handle, RESYNC_FAILED_INTERVAL);
+		}
+
+		return;
+	}
+
+	if (status == DNS_EAI_INPROGRESS && info != NULL) {
+		switch (info->ai_family) {
+		case NET_AF_INET:
+			if (sntp_addrlen > 0) {
+				/* Already got address, skip others. */
+				break;
+			}
+
+			sntp_addrlen = info->ai_addrlen;
+			sntp_addr.ss_family = NET_AF_INET;
+			net_ipv4_addr_copy_raw(net_sin(net_sad(&sntp_addr))->sin_addr.s4_addr,
+					       net_sin(&info->ai_addr)->sin_addr.s4_addr);
+			net_sin(net_sad(&sntp_addr))->sin_port = net_htons(123);
+			break;
+
+		case NET_AF_INET6:
+			if (sntp_addrlen > 0) {
+				/* Already got address, skip others. */
+				break;
+			}
+
+			sntp_addrlen = info->ai_addrlen;
+			sntp_addr.ss_family = NET_AF_INET6;
+			net_ipv6_addr_copy_raw(net_sin6(net_sad(&sntp_addr))->sin6_addr.s6_addr,
+					       net_sin6(&info->ai_addr)->sin6_addr.s6_addr);
+			net_sin6(net_sad(&sntp_addr))->sin6_port = net_htons(123);
+			break;
+
+		default:
+			/* Ignore other records. */
+			break;
+		}
+	}
+}
+
+static int dns_query_async(void)
+{
+	enum dns_query_type type;
+	int ret;
+
+	memset(&sntp_addr, 0, sizeof(sntp_addr));
+	sntp_addrlen = 0;
+
+	if (IS_ENABLED(CONFIG_NET_IPV4)) {
+		sntp_addr.ss_family = NET_AF_INET;
+		type = DNS_QUERY_TYPE_A;
+	} else {
+		sntp_addr.ss_family = NET_AF_INET6;
+		type = DNS_QUERY_TYPE_AAAA;
+	}
+
+	ret = dns_get_addr_info(CONFIG_NET_CONFIG_SNTP_INIT_SERVER,
+				type, NULL, dns_result_cb,
+				NULL, CONFIG_NET_CONFIG_SNTP_INIT_TIMEOUT);
+	if (ret < 0) {
+		goto out;
+	}
+
+out:
+	return ret;
+}
+
 static void sntp_resync_handler(struct k_work *work)
 {
+	int ret;
+
 	ARG_UNUSED(work);
-	(void)net_init_clock_via_sntp();
+
+#ifdef CONFIG_NET_CONFIG_SNTP_INIT_SERVER_USE_DHCPV4_OPTION
+	struct net_if *iface = net_if_get_default();
+
+	if (!net_ipv4_is_addr_unspecified(&iface->config.dhcpv4.ntp_addr)) {
+		sntp_addrlen = sizeof(struct net_sockaddr_in);
+		sntp_addr.ss_family = NET_AF_INET;
+		net_ipv4_addr_copy_raw(net_sin(net_sad(&sntp_addr))->sin_addr.s4_addr,
+					       iface->config.dhcpv4.ntp_addr.s4_addr);
+		net_sin(net_sad(&sntp_addr))->sin_port = net_htons(123);
+
+		ret = sntp_query_async(net_sad(&sntp_addr), sntp_addrlen);
+		if (ret < 0) {
+			LOG_ERR("Cannot set time using SNTP: %d", ret);
+		}
+
+		goto out;
+	}
+
+	if (sizeof(CONFIG_NET_CONFIG_SNTP_INIT_SERVER) == 1) {
+		/* Empty address, skip using SNTP via Kconfig defaults */
+		ret = -EINVAL;
+		goto out;
+	}
+#endif
+
+	ret = dns_query_async();
+	if (ret < 0) {
+		LOG_ERR("Failed to initiate DNS query for SNTP server (%d)", ret);
+		goto out;
+	}
+
+out:
+	k_work_reschedule(
+		&sntp_resync_work_handle,
+		(ret < 0) ? K_SECONDS(CONFIG_NET_CONFIG_SNTP_INIT_RESYNC_ON_FAILURE_INTERVAL)
+			  : K_SECONDS(CONFIG_NET_CONFIG_SNTP_INIT_RESYNC_INTERVAL));
```

Comment:
The regular `net_init_clock_via_sntp` is no longer used, when CONFIG_NET_CONFIG_SNTP_INIT_USE_CONNECTION_MANAGER is enabled, as it uses the resync variant always.